File: Client/RelaxKonOS.Client.Android/app/src/main/java/app/relaxkonos/mobile/servercenter/ServerCenterDeploymentClient.kt
package app.relaxkonos.mobile.servercenter

import java.io.ByteArrayInputStream
import java.io.File
import java.io.IOException
import java.io.InputStream

enum class ServerHostPlatform { Linux, Windows }

/** A source that can be reopened for validation and SFTP upload. */
class ServerCenterUploadAsset(
    val contentLength: Long?,
    private val openStream: () -> InputStream,
) {
    fun open(): InputStream = openStream()

    companion object {
        fun bytes(value: ByteArray): ServerCenterUploadAsset =
            ServerCenterUploadAsset(value.size.toLong()) { ByteArrayInputStream(value) }

        fun file(value: File): ServerCenterUploadAsset =
            ServerCenterUploadAsset(value.length()) { value.inputStream() }
    }
}

/** The remote staging identity retained until the authoritative operation receipt is read. */
data class ServerCenterStagedOperation(
    val operationId: String,
    val platform: ServerHostPlatform,
    val remoteDirectory: String,
)

/**
 * Android deployment operation layer. It binds local signed-release verification to the built-in
 * SSH/SFTP transport and invokes only the fixed launcher actions.
 */
class ServerCenterDeploymentClient(private val transport: ServerCenterSshTransport) {

    suspend fun stage(
        request: ServerDeploymentRequest,
        platform: ServerHostPlatform,
        launcher: ServerCenterUploadAsset,
        verifier: ServerCenterUploadAsset,
        signedArchive: File? = null,
        expectedRuntime: ServerRuntimeIdentifier? = null,
        trustedKeyId: String? = null,
        trustedPublicKeyPem: String? = null,
        uploadProgress: ((Double) -> Unit)? = null,
    ): ServerCenterStagedOperation {
        check(transport.isConnected) { "A trusted SSH session is required." }
        val requestBytes = ServerDeploymentWire.writeRequest(request)
        val needsArchive = request.kind == ServerDeploymentKind.Install || request.kind == ServerDeploymentKind.Upgrade
        if (needsArchive) {
            val options = requireNotNull(request.options) { "Deployment options are required." }
            val archive = requireNotNull(signedArchive) { "A signed release is required." }
            val runtime = requireNotNull(expectedRuntime) { "The expected release RID is required." }
            val keyId = requireNotNull(trustedKeyId?.takeIf { it.isNotBlank() }) { "A trusted key ID is required." }
            val publicKey = requireNotNull(trustedPublicKeyPem?.takeIf { it.isNotBlank() }) {
                "A trusted public key is required."
            }
            require(archive.isFile && TRUSTED_KEY_ID.matches(keyId) &&
                ServerDeploymentInputRules.isSafeStagedPackageName(options.stagedPackageName) &&
                ServerDeploymentInputRules.isSha256(options.packageDigest)) {
                "A signed, staged release and a trusted key are required."
            }
            require(modeMatchesPlatform(options.mode, platform)) { "Installation mode does not match the host platform." }
            require(runtimeMatchesPlatform(runtime, platform)) { "Release RID does not match the host platform." }

            val expectedKind = if (options.mode == ServerInstallMode.LinuxUser) {
                ServerReleasePackageKind.UserServer
            } else ServerReleasePackageKind.Server
            val checkedRelease = ServerReleaseArchiveVerifier.verify(
                archiveFile = archive,
                expectedKind = expectedKind,
                expectedRuntime = runtime,
                trustedPublicKeys = mapOf(keyId to publicKey),
                expectedArchiveSha256 = options.packageDigest,
            )
            if (!checkedRelease.verified) {
                throw IOException("${checkedRelease.problemCode}: release verification failed before upload.")
            }
        }

        val directory = createPrivateDirectory(platform)
        val staged = ServerCenterStagedOperation(request.operationId.lowercase(), platform, directory)
        upload(verifier, staged, if (platform == ServerHostPlatform.Windows) "release-verifier.exe" else "release-verifier")
        upload(launcher, staged, if (platform == ServerHostPlatform.Windows) "RelaxKonOS-Deploy.ps1" else "relaxkonos-deploy.sh")
        if (needsArchive) {
            upload(
                ServerCenterUploadAsset.bytes(trustedPublicKeyPem!!.toByteArray(Charsets.UTF_8)),
                staged,
                "release-public.pem",
            )
            upload(
                ServerCenterUploadAsset.bytes(trustedKeyId!!.toByteArray(Charsets.US_ASCII)),
                staged,
                "release-key-id.txt",
            )
            upload(
                ServerCenterUploadAsset.file(signedArchive!!),
                staged,
                request.options!!.stagedPackageName!!,
                uploadProgress,
            )
        }
        upload(ServerCenterUploadAsset.bytes(requestBytes), staged, "request.json")
        if (platform == ServerHostPlatform.Linux) {
            val chmod = transport.run(
                "chmod 700 '$directory/release-verifier' '$directory/relaxkonos-deploy.sh'",
            )
            if (!chmod.succeeded) throw IOException("Unable to mark the staged deployment tools executable.")
        }
        return staged
    }

    suspend fun execute(staged: ServerCenterStagedOperation): ServerDeploymentOperation {
        transport.run(launcherCommand(staged, query = false))
        // SSH exit status is not proof of success. The persistent receipt is authoritative.
        return query(staged)
    }

    suspend fun query(staged: ServerCenterStagedOperation): ServerDeploymentOperation {
        val result = transport.run(launcherCommand(staged, query = true))
        if (!result.succeeded) throw IOException("The remote operation receipt could not be read.")
        return ServerDeploymentWire.readOperation(result.standardOutput.trim(), staged.operationId)
    }

    private suspend fun createPrivateDirectory(platform: ServerHostPlatform): String {
        val command = when (platform) {
            ServerHostPlatform.Linux -> "umask 077; mktemp -d -p /tmp relaxkonos-deploy.XXXXXXXX"
            ServerHostPlatform.Windows -> {
                val script = "\$p=Join-Path \$env:TEMP ('relaxkonos-deploy-'+[guid]::NewGuid().ToString('N'));" +
                    "[IO.Directory]::CreateDirectory(\$p)|Out-Null;" +
                    "\$u=[Security.Principal.WindowsIdentity]::GetCurrent().Name;" +
                    "& icacls \$p /inheritance:r /grant:r (\$u+':(OI)(CI)F') 'SYSTEM:(OI)(CI)F' " +
                    "'Administrators:(OI)(CI)F' *> \$null;if(\$LASTEXITCODE -ne 0){exit 1};" +
                    "[Console]::WriteLine(\$p)"
                "powershell.exe -NoProfile -NonInteractive -EncodedCommand " +
                    Base64Codec.encode(script.toByteArray(Charsets.UTF_16LE))
            }
        }
        val result = transport.run(command)
        if (!result.succeeded) throw IOException("The remote private staging directory could not be created.")
        val path = result.standardOutput.trim()
        val valid = '\n' !in path && '\r' !in path && when (platform) {
            ServerHostPlatform.Linux -> LINUX_STAGING_PATH.matches(path)
            ServerHostPlatform.Windows -> WINDOWS_STAGING_PATH.matches(path)
        }
        if (!valid) throw IOException("The remote staging path is invalid.")
        return path
    }

    private suspend fun upload(
        source: ServerCenterUploadAsset,
        staged: ServerCenterStagedOperation,
        fileName: String,
        progress: ((Double) -> Unit)? = null,
    ) {
        source.open().use { content ->
            transport.upload(content, source.contentLength, remotePath(staged, fileName), progress)
        }
    }

    private fun launcherCommand(staged: ServerCenterStagedOperation, query: Boolean): String {
        if (staged.platform == ServerHostPlatform.Linux) {
            require(LINUX_STAGING_PATH.matches(staged.remoteDirectory)) { "Invalid staging directory." }
            val action = if (query) " --query ${staged.operationId}" else " --run"
            return "bash '${staged.remoteDirectory}/relaxkonos-deploy.sh'$action"
        }
        require(WINDOWS_STAGING_PATH.matches(staged.remoteDirectory)) { "Invalid staging directory." }
        val scriptPath = staged.remoteDirectory.trimEnd('\\', '/') + "\\RelaxKonOS-Deploy.ps1"
        val command = "& '${scriptPath.replace("'", "''")}'" +
            if (query) " -QueryOperationId '${staged.operationId}'" else ""
        return "powershell.exe -NoProfile -NonInteractive -EncodedCommand " +
            Base64Codec.encode(command.toByteArray(Charsets.UTF_16LE))
    }

    private fun remotePath(staged: ServerCenterStagedOperation, fileName: String): String =
        staged.remoteDirectory.trimEnd('/', '\\').replace('\\', '/') + "/" + fileName

    private fun modeMatchesPlatform(mode: ServerInstallMode?, platform: ServerHostPlatform): Boolean = when (platform) {
        ServerHostPlatform.Linux -> mode == ServerInstallMode.LinuxSystem || mode == ServerInstallMode.LinuxUser
        ServerHostPlatform.Windows -> mode == ServerInstallMode.WindowsSystem
    }

    private fun runtimeMatchesPlatform(runtime: ServerRuntimeIdentifier, platform: ServerHostPlatform): Boolean =
        when (platform) {
            ServerHostPlatform.Linux -> runtime == ServerRuntimeIdentifier.LinuxX64 ||
                runtime == ServerRuntimeIdentifier.LinuxArm64
            ServerHostPlatform.Windows -> runtime == ServerRuntimeIdentifier.WinX64 ||
                runtime == ServerRuntimeIdentifier.WinArm64
        }

    private companion object {
        val TRUSTED_KEY_ID = Regex("^[A-Za-z0-9._-]{1,128}$")
        val LINUX_STAGING_PATH = Regex("^/tmp/relaxkonos-deploy\\.[A-Za-z0-9]{8,32}$")
        val WINDOWS_STAGING_PATH = Regex("^[A-Za-z]:[\\\\/].*[\\\\/]relaxkonos-deploy-[0-9a-f]{32}$")
    }
}

File: Client/RelaxKonOS.Client/Services/ServerCenter/ServerCenterDeploymentClient.cs
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;
using RelaxKonOS.Protocol.Common;
using RelaxKonOS.Protocol.ServerCenter;

namespace RelaxKonOS.Client.Services.ServerCenter;

/// <summary>Remote staging identity. Keep this with the operation id until the authoritative receipt is read.</summary>
public sealed record ServerCenterStagedOperation(Guid OperationId, HostPlatformKind Platform, string RemoteDirectory);

/// <summary>
/// Binds the signed release verifier to the built-in SSH/SFTP transport. The caller supplies trusted
/// release assets from its pinned release configuration, never from the archive being uploaded.
/// </summary>
public sealed class ServerCenterDeploymentClient(IServerCenterSshTransport transport)
{
    private static readonly Regex LinuxStagingPath = new(
        @"^/tmp/relaxkonos-deploy\.[A-Za-z0-9]{8,32}$", RegexOptions.CultureInvariant);
    private static readonly Regex WindowsStagingPath = new(
        @"^[A-Za-z]:[\\/].*[\\/]relaxkonos-deploy-[0-9a-f]{32}$", RegexOptions.CultureInvariant);

    public async Task<ServerCenterStagedOperation> StageAsync(
        ServerDeploymentRequest request,
        HostPlatformKind platform,
        Stream launcher,
        Stream verifier,
        Stream? signedArchive,
        ServerRuntimeIdentifier? expectedRuntime,
        string? trustedKeyId,
        string? trustedPublicKeyPem,
        CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);
        ArgumentNullException.ThrowIfNull(launcher);
        ArgumentNullException.ThrowIfNull(verifier);
        if (!transport.IsConnected) throw new InvalidOperationException("A trusted SSH session is required.");
        if (!launcher.CanRead || !verifier.CanRead || !launcher.CanSeek || !verifier.CanSeek)
            throw new ArgumentException("Seekable launcher and verifier streams are required.");
        if (request.SchemaVersion != ServerDeploymentProtocol.Version || request.OperationId == Guid.Empty)
            throw new ArgumentException("The deployment request is invalid.", nameof(request));

        var requestBytes = JsonSerializer.SerializeToUtf8Bytes(request, RelaxKonOSJsonOptions.Default);
        if (!ServerDeploymentRequestWireValidation.IsStrictRequest(requestBytes))
            throw new ArgumentException("The deployment request is not a strict wire request.", nameof(request));

        var needsArchive = request.Kind is ServerDeploymentKind.Install or ServerDeploymentKind.Upgrade;
        if (needsArchive)
        {
            if (signedArchive is null || !signedArchive.CanSeek ||
                expectedRuntime is null || string.IsNullOrWhiteSpace(trustedKeyId) ||
                string.IsNullOrWhiteSpace(trustedPublicKeyPem) ||
                !Regex.IsMatch(trustedKeyId, @"^[A-Za-z0-9._-]{1,128}$", RegexOptions.CultureInvariant) ||
                !ServerDeploymentInputRules.IsSafeStagedPackageName(request.Options?.StagedPackageName) ||
                !ServerDeploymentInputRules.IsSha256(request.Options?.PackageDigest))
                throw new ArgumentException("A signed, staged release and a trusted key are required.");
            if (platform == HostPlatformKind.Windows && request.Options?.Mode != ServerInstallMode.WindowsSystem ||
                platform == HostPlatformKind.Linux && request.Options?.Mode is not
                    (ServerInstallMode.LinuxSystem or ServerInstallMode.LinuxUser))
                throw new ArgumentException("The installation mode does not match the host platform.");
            if (platform == HostPlatformKind.Windows !=
                (expectedRuntime is ServerRuntimeIdentifier.WinX64 or ServerRuntimeIdentifier.WinArm64))
                throw new ArgumentException("The release RID does not match the host platform.");

            var kind = request.Options!.Mode == ServerInstallMode.LinuxUser
                ? ServerReleasePackageKind.UserServer : ServerReleasePackageKind.Server;
            var keys = new Dictionary<string, string>(StringComparer.Ordinal) { [trustedKeyId] = trustedPublicKeyPem };
            var checkedRelease = ServerReleaseArchiveVerifier.Verify(
                signedArchive, kind, expectedRuntime.Value, keys, request.Options.PackageDigest);
            if (!checkedRelease.Verified)
                throw new InvalidDataException($"{checkedRelease.ProblemCode}: release verification failed before upload.");
        }

        var directory = await CreatePrivateDirectoryAsync(platform, cancellationToken).ConfigureAwait(false);
        var staged = new ServerCenterStagedOperation(request.OperationId, platform, directory);
        await UploadFromStartAsync(verifier, staged, platform == HostPlatformKind.Windows
            ? "release-verifier.exe" : "release-verifier", cancellationToken).ConfigureAwait(false);
        await UploadFromStartAsync(launcher, staged, platform == HostPlatformKind.Windows
            ? "RelaxKonOS-Deploy.ps1" : "relaxkonos-deploy.sh", cancellationToken).ConfigureAwait(false);
        if (needsArchive)
        {
            await UploadBytesAsync(Encoding.UTF8.GetBytes(trustedPublicKeyPem!), staged, "release-public.pem", cancellationToken)
                .ConfigureAwait(false);
            await UploadBytesAsync(Encoding.ASCII.GetBytes(trustedKeyId!), staged, "release-key-id.txt", cancellationToken)
                .ConfigureAwait(false);
            await UploadFromStartAsync(signedArchive!, staged, request.Options!.StagedPackageName!, cancellationToken)
                .ConfigureAwait(false);
        }
        await UploadBytesAsync(requestBytes, staged, "request.json", cancellationToken).ConfigureAwait(false);
        if (platform == HostPlatformKind.Linux)
        {
            var chmod = await transport.RunAsync("chmod 700 '" + directory + "/release-verifier' '" +
                directory + "/relaxkonos-deploy.sh'", cancellationToken).ConfigureAwait(false);
            if (!chmod.Succeeded) throw new IOException("Unable to mark the staged deployment tools executable.");
        }
        return staged;
    }

    /// <summary>
    /// Stages only the fixed query tools for reconnecting to an existing operation receipt.  It does
    /// not create a deployment request, upload a package, or start a new operation: the supplied ID
    /// is solely the immutable key of a previously recorded remote operation.
    /// </summary>
    public async Task<ServerCenterStagedOperation> StageQueryAsync(
        Guid operationId,
        HostPlatformKind platform,
        Stream launcher,
        Stream verifier,
        CancellationToken cancellationToken)
    {
        if (operationId == Guid.Empty) throw new ArgumentException("An operation id is required.", nameof(operationId));
        ArgumentNullException.ThrowIfNull(launcher);
        ArgumentNullException.ThrowIfNull(verifier);
        if (!transport.IsConnected) throw new InvalidOperationException("A trusted SSH session is required.");
        if (!launcher.CanRead || !verifier.CanRead || !launcher.CanSeek || !verifier.CanSeek)
            throw new ArgumentException("Seekable launcher and verifier streams are required.");

        var directory = await CreatePrivateDirectoryAsync(platform, cancellationToken).ConfigureAwait(false);
        var staged = new ServerCenterStagedOperation(operationId, platform, directory);
        await UploadFromStartAsync(verifier, staged, platform == HostPlatformKind.Windows
            ? "release-verifier.exe" : "release-verifier", cancellationToken).ConfigureAwait(false);
        await UploadFromStartAsync(launcher, staged, platform == HostPlatformKind.Windows
            ? "RelaxKonOS-Deploy.ps1" : "relaxkonos-deploy.sh", cancellationToken).ConfigureAwait(false);
        if (platform == HostPlatformKind.Linux)
        {
            var chmod = await transport.RunAsync("chmod 700 '" + directory + "/release-verifier' '" +
                directory + "/relaxkonos-deploy.sh'", cancellationToken).ConfigureAwait(false);
            if (!chmod.Succeeded) throw new IOException("Unable to mark the staged deployment tools executable.");
        }
        return staged;
    }

    public async Task<ServerDeploymentOperationDto> ExecuteAsync(
        ServerCenterStagedOperation staged, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(staged);
        await transport.RunAsync(LauncherCommand(staged, query: false), cancellationToken)
            .ConfigureAwait(false);
        // An SSH exit status is not proof of success. Query the persistent receipt even on a nonzero exit.
        return await QueryAsync(staged, cancellationToken).ConfigureAwait(false);
    }

    public async Task<ServerDeploymentOperationDto> QueryAsync(
        ServerCenterStagedOperation staged, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(staged);
        var result = await transport.RunAsync(LauncherCommand(staged, query: true), cancellationToken)
            .ConfigureAwait(false);
        if (!result.Succeeded) throw new IOException("The remote operation receipt could not be read.");
        return ServerDeploymentRecordReader.ReadRecord(result.StandardOutput.Trim(), staged.OperationId);
    }

    private async Task<string> CreatePrivateDirectoryAsync(HostPlatformKind platform, CancellationToken cancellationToken)
    {
        string command;
        if (platform == HostPlatformKind.Linux)
        {
            command = "umask 077; mktemp -d -p /tmp relaxkonos-deploy.XXXXXXXX";
        }
        else if (platform == HostPlatformKind.Windows)
        {
            const string script = "$p=Join-Path $env:TEMP ('relaxkonos-deploy-'+[guid]::NewGuid().ToString('N'));" +
                "[IO.Directory]::CreateDirectory($p)|Out-Null;" +
                "$u=[Security.Principal.WindowsIdentity]::GetCurrent().Name;" +
                "& icacls $p /inheritance:r /grant:r ($u+':(OI)(CI)F') 'SYSTEM:(OI)(CI)F' 'Administrators:(OI)(CI)F' *> $null;" +
                "if($LASTEXITCODE -ne 0){exit 1};[Console]::WriteLine($p)";
            command = "powershell.exe -NoProfile -NonInteractive -EncodedCommand " +
                Convert.ToBase64String(Encoding.Unicode.GetBytes(script));
        }
        else throw new ArgumentOutOfRangeException(nameof(platform));

        var result = await transport.RunAsync(command, cancellationToken).ConfigureAwait(false);
        if (!result.Succeeded) throw new IOException("The remote private staging directory could not be created.");
        var path = result.StandardOutput.Trim();
        if (path.Contains('\n') || path.Contains('\r') ||
            !(platform == HostPlatformKind.Linux ? LinuxStagingPath : WindowsStagingPath).IsMatch(path))
            throw new InvalidDataException("The remote staging path is invalid.");
        return path;
    }

    private async Task UploadFromStartAsync(Stream source, ServerCenterStagedOperation staged,
        string fileName, CancellationToken cancellationToken)
    {
        source.Position = 0;
        await transport.UploadAsync(source, RemotePath(staged, fileName), null, cancellationToken)
            .ConfigureAwait(false);
    }

    private async Task UploadBytesAsync(byte[] bytes, ServerCenterStagedOperation staged,
        string fileName, CancellationToken cancellationToken)
    {
        using var source = new MemoryStream(bytes, writable: false);
        await transport.UploadAsync(source, RemotePath(staged, fileName), null, cancellationToken)
            .ConfigureAwait(false);
    }

    private static string RemotePath(ServerCenterStagedOperation staged, string fileName) =>
        staged.RemoteDirectory.TrimEnd('/', '\\').Replace('\\', '/') + '/' + fileName;

    private static string LauncherCommand(ServerCenterStagedOperation staged, bool query)
    {
        if (staged.Platform == HostPlatformKind.Linux)
        {
            if (!LinuxStagingPath.IsMatch(staged.RemoteDirectory)) throw new ArgumentException("Invalid staging directory.");
            var action = query ? " --query " + staged.OperationId.ToString("D") : " --run";
            return "bash '" + staged.RemoteDirectory + "/relaxkonos-deploy.sh'" + action;
        }
        if (staged.Platform != HostPlatformKind.Windows || !WindowsStagingPath.IsMatch(staged.RemoteDirectory))
            throw new ArgumentException("Invalid staging directory.");
        var scriptPath = staged.RemoteDirectory.TrimEnd('\\', '/') + "\\RelaxKonOS-Deploy.ps1";
        var escapedPath = scriptPath.Replace("'", "''", StringComparison.Ordinal);
        var command = "& '" + escapedPath + "'" +
            (query ? " -QueryOperationId '" + staged.OperationId.ToString("D") + "'" : "");
        return "powershell.exe -NoProfile -NonInteractive -EncodedCommand " +
            Convert.ToBase64String(Encoding.Unicode.GetBytes(command));
    }
}
